src/analysis/processing/qgsalgorithmfillnodata.cpp

```diff
@@ -0,0 +1,158 @@
+}
+
+QString QgsFillNoDataAlgorithm::group() const
+{
+  return QObject::tr( "Raster tools" );
+}
+
+QString QgsFillNoDataAlgorithm::groupId() const
+{
+  return QStringLiteral( "rastertools" );
+}
+
+void QgsFillNoDataAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterRasterLayer( QStringLiteral( "INPUT" ), QStringLiteral( "Raster input" ) ) );
+  addParameter( new QgsProcessingParameterBand( QStringLiteral( "BAND" ), QObject::tr( "Band Number" ), 1, QStringLiteral( "INPUT" ) ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "FILL_VALUE" ), QObject::tr( "Fill value" ), QgsProcessingParameterNumber::Double, 1, false, 1, 10000000 ) );
+  addParameter( new QgsProcessingParameterRasterDestination( QStringLiteral( "OUTPUT" ), QObject::tr( "Output raster" ) ) );
+}
+
+QString QgsFillNoDataAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm resets the NoData values in the input raster "
+                      "to a chosen value. This value can be set by the user using the Fill value parameter. "
+                      "The algorithm respects the input raster data type (eg. a fill value with commas gets truncated "
+                      "when applied to an integer raster)." );
+}
+
+QgsFillNoDataAlgorithm *QgsFillNoDataAlgorithm::createInstance() const
+{
+  return new QgsFillNoDataAlgorithm();
+}
+
+bool QgsFillNoDataAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  Q_UNUSED( feedback );
+  mInputRaster = parameterAsRasterLayer( parameters, QStringLiteral( "INPUT" ), context );
+  mFillValue = parameterAsDouble( parameters, QStringLiteral( "FILL_VALUE" ), context );
+
+  if ( !mInputRaster )
+    throw QgsProcessingException( invalidRasterError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  mBand = parameterAsInt( parameters, QStringLiteral( "BAND" ), context );
+  if ( mBand < 1 || mBand > mInputRaster->bandCount() )
+    throw QgsProcessingException( QObject::tr( "Invalid band number for BAND (%1): Valid values for input raster are 1 to %2" ).arg( mBand ).arg( mInputRaster->bandCount() ) );
+
+  mInterface.reset( mInputRaster->dataProvider()->clone() );
+  mInputNoDataValue = mInputRaster->dataProvider()->sourceNoDataValue( mBand );
+  mExtent = mInputRaster->extent();
+  mLayerWidth = mInputRaster->width();
+  mLayerHeight = mInputRaster->height();
+  mCrs = mInputRaster->crs();
+  mNbCellsXProvider = mInterface->xSize();
+  mNbCellsYProvider = mInterface->ySize();
+  return true;
+}
+
+QVariantMap QgsFillNoDataAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  //test if input dataset has NoData
+  if ( !mInputRaster->dataProvider()->sourceHasNoDataValue( mBand ) )
+    throw QgsProcessingException( QObject::tr( "Input raster has no NoData values. There exist no NoData cells to fill." ) );
```

Comment:
I'm not sure we want this to be an exception here, and instead think we should just raise a warning (`feedback->reportError( ... )`). If we raise an exception it will terminate a model when encountered, and potentially users may need a model which can gracefully handle inputs either with or without nodata values present.